Improve Player Lifecycle Docs (#869)
spatialos/gdk-for-unity: docs/modules/player-lifecycle/basic-player-creation.md

<%(TOC)%>
# Implementing basic player creation

<%(Callout message="
Before reading this document, make sure you have read the following documentation on:

* [Player Lifecycle Feature Module]({{urlRoot}}/modules/player-lifecycle/overview)
* [Creating entity templates]({{urlRoot}}/reference/concepts/entity-templates)
* [Workers in the GDK]({{urlRoot}}/reference/concepts/worker)
")%>

> By default, the module sends a player creation request as soon as the client-worker instance connects to SpatialOS. The server-worker instance which receives the request spawns a [SpatialOS entity]({{urlRoot}}/reference/glossary#spatialos-entity) to represent the player. It then deletes the player entity after multiple consecutive unsuccessful [heartbeats]({{urlRoot}}/modules/player-lifecycle/heartbeating).

Reviewer: ```suggestion
By default, the module sends a player creation request as soon as the client-worker instance connects to SpatialOS. The server-worker instance receiving the request spawns a [SpatialOS entity]({{urlRoot}}/reference/glossary#spatialos-entity) to represent the player.The player entity gets deleted, if the server-worker fails to receive multiple [heartbeats]({{urlRoot}}/modules/player-lifecycle/heartbeating) consecutively.
```

<%(TOC)%>
# Implementing basic player creation

<%(Callout message="
Before reading this document, make sure you have read the following documentation on:

* [Player Lifecycle Feature Module]({{urlRoot}}/modules/player-lifecycle/overview)
* [Creating entity templates]({{urlRoot}}/reference/concepts/entity-templates)
* [Workers in the GDK]({{urlRoot}}/reference/concepts/worker)
")%>

> By default, the module sends a player creation request as soon as the client-worker instance connects to SpatialOS. The server-worker instance receiving the request spawns a [SpatialOS entity]({{urlRoot}}/reference/glossary#spatialos-entity) to represent the player.The player entity gets deleted, if the server-worker fails to receive multiple [heartbeats]({{urlRoot}}/modules/player-lifecycle/heartbeating) consecutively.

Author: Maybe it's worth an expandable? We say how to create players but never how they get deleted

> By default, the module sends a player creation request as soon as the client-worker instance connects to SpatialOS. The server-worker instance receiving the request spawns a [SpatialOS entity]({{urlRoot}}/reference/glossary#spatialos-entity) to represent the player.The player entity gets deleted, if the server-worker fails to receive multiple [heartbeats]({{urlRoot}}/modules/player-lifecycle/heartbeating) consecutively.

Reviewer 2: I just feel like someone who is reading this page is only concerned with getting a player spawning - the heartbeating page covers how they get deleted

<%(TOC)%>
# Implementing basic player creation

<%(Callout message="
Before reading this document, make sure you have read the following documentation on:

* [Player Lifecycle Feature Module]({{urlRoot}}/modules/player-lifecycle/overview)
* [Creating entity templates]({{urlRoot}}/reference/concepts/entity-templates)
* [Workers in the GDK]({{urlRoot}}/reference/concepts/worker)
")%>

> By default, the module automatically sends a player creation request as soon as the client-worker instance connects to SpatialOS. This spawns a [SpatialOS entity]({{urlRoot}}/reference/glossary#spatialos-entity) representing a player.

Reviewer 2: ~automatically~

Reviewer 2: ```
The server-workers instance which receives the request spawns a [SpatialOS entity]({{urlRoot}}/reference/glossary#spatialos-entity) to represent the player.
```

To do this, you need to:

1. Set up your worker connector.
1. Define the entity template for your player entities.

> 1. Configure what entity template to use for player creation.

Reviewer 2: I'm not a huge fan of this preamble for a short number of steps. I defer to @jessicafalk or @ElleEss on what our policy is for this. 

Also saying `To do this` after saying it automatically spawns a player feels weird

Reviewer: usually I do like lists to emphasize of what's gonna happen next, but in this case I agree with Jamie.  Can get rid of this, the steps are really short :)

> ## Set up your worker connector

Reviewer 2: Would be nice to keep this consistent across every feature module documentation

Author: I tried to mimic the worker connector bit from the gameobject creation module docs

Reviewer 2: I think its clearer if we don't have bullets and inlined code for the snippets, i.e. - 

---

You need to add the underlying systems to your worker. Open your [`WorkerConnector` implementations]({{urlRoot}}/reference/workflows/monobehaviour/creating-workers) and add one of the following lines to the `HandleWorkerConnectionEstablished` method.

**If this is a client-worker:**

```csharp
    TransformSynchronizationHelper.AddClientSystems(Worker.World);
```

**If this is a server-worker:**

```csharp
    TransformSynchronizationHelper.AddServerWorkers(Worker.World);
```

Author: That makes sense, and thanks for the neat example 

will update

After setting up the module, you must enable it by adding the necessary systems to your workers. This is done in the `HandleWorkerConnectionEstablished()` method of your [`WorkerConnector`]({{urlRoot}}/reference/workflows/monobehaviour/creating-workers), by using the following code snippets:

1. On a [client-worker]({{urlRoot}}/reference/glossary#client-worker): `PlayerLifecycleHelper.AddClientSystems(Worker.World)`
1. On a [server-worker]({{urlRoot}}/reference/glossary#server-worker): `PlayerLifecycleHelper.AddServerSystems(Worker.World)`

To change this behaviour, read the documentation on [custom player creation]({{urlRoot}}/modules/player-lifecycle/custom-player-creation).

## Define the player entity template

> The module takes care of spawning the player entity as soon as a client-worker connects. To enable this behaviour, the server-worker responsible for handling requests to spawn player entities needs to know which [entity template]({{urlRoot}}/reference/concepts/entity-templates) to use when sending the entity creation request to the [SpatialOS Runtime]({{urlRoot}}/reference/glossary#spatialos-runtime).

Reviewer 2: ~The module takes care of spawning the player entity as soon as a client-worker connects. To enable this behaviour,~

Create a method that returns an `EntityTemplate` object and takes the following parameters to define your player [entity template]({{urlRoot}}/reference/concepts/entity-templates):

* `string workerId`: The ID of the worker that wants to spawn this player entity.
*  `byte[] playerCreationArguments`: a serialized byte array of arguments provided by the worker instance sending the player creation request.

When defining the entity template, you need to use the `AddPlayerLifecycleComponents` method. This method adds lifecycle SpatialOS components to the player entity template; these components are necessary to manage the player lifecycle for this entity.

The following code snippet shows an example on how to implement such a method:

```csharp
public static class PlayerTemplate
{
    public static EntityTemplate CreatePlayerEntityTemplate(string workerId, byte[] playerCreationArguments)
    {
        // Obtain unique client attribute of the client-worker that requested the player entity
        var clientAttribute = EntityTemplate.GetWorkerAccessAttribute(workerId);
        // Obtain the attribute of your server-worker

>         var serverAttribute = "UnityGameLogic";

Reviewer 2: Unnecessary line

        var entityTemplate = new EntityTemplate();
        entityTemplate.AddPosition(new Position.Snapshot(new Coordinates()), serverAttribute);
        // add all components that you want the player entity to have
        AddPlayerLifecycleComponents(entityTemplate, workerId, serverAttribute);

        return entityTemplate;
    }
}
```

> ## Configure an entity template to use for player creation

Reviewer 2: This sounds suspiciously similar to the heading above

Configure the entity template delegate?

You need to configure the `PlayerLifecycleConfig` configuration class to use the player entity template. Do this by setting the `PlayerLifecycleConfig.CreatePlayerEntityTemplate` field to `PlayerTemplate.CreatePlayerEntityTemplate`.

The following example is a code snippet using a script that runs once when starting the game to initialize the configuration for the player lifecycle module:

```csharp
public static class OneTimeInitialization
{
    private static bool initialized;

    // Using this attribute will trigger the Init() method whenever a Scene gets loaded.
    [RuntimeInitializeOnLoadMethod]
    private static void Init()
    {
        if (initialized)
        {
            // ensures to only run it the first time a Scene gets loaded.
            return;
        }

        initialized = true;
        // initialize your game...
        // ...
        // Setup template to use for player on connecting client
        PlayerLifecycleConfig.CreatePlayerEntityTemplate = PlayerTemplate.CreatePlayerEntityTemplate;
    }
}
```

> ## When is a player entity deleted?

Reviewer 2: This feels out of place here.

We can point users to the heartbeating _or_ custom player creation docs but probably not worth talking in too much detail

**If this is a client-worker:**

```csharp
    PlayerLifecycleHelper.AddClientSystems(Worker.World);
```

**If this is a server-worker:**

```csharp
    PlayerLifecycleHelper.AddServerWorkers(Worker.World);
```

To change this behaviour, read the documentation on [custom player creation]({{urlRoot}}/modules/player-lifecycle/custom-player-creation).

## Define the player entity template

The server-worker responsible for handling requests to spawn player entities needs to know which [entity template]({{urlRoot}}/reference/concepts/entity-templates) to use when sending the entity creation request to the [SpatialOS Runtime]({{urlRoot}}/reference/glossary#spatialos-runtime).

Create a method that returns an `EntityTemplate` object and takes the following parameters to define your player [entity template]({{urlRoot}}/reference/concepts/entity-templates):

* `string workerId`: The ID of the worker that wants to spawn this player entity.
*  `byte[] playerCreationArguments`: a serialized byte array of arguments provided by the worker instance sending the player creation request.

> When defining the entity template, you need to use the `AddPlayerLifecycleComponents` method. This method adds lifecycle SpatialOS components to the player entity template; these components are necessary to manage the player lifecycle for this entity.

Reviewer: ```suggestion
When defining the entity template, you need to use the `AddPlayerLifecycleComponents` method. This method adds SpatialOS components to the player entity template that are used by the Player Lifecycle Feature Module to manage the lifecycle of this entity.

Reviewer: not sure which one is correct: of vs for
